x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/GeometricShapeSyntheticSourceSupport.java

```diff
@@ -92,48 +95,42 @@ private Value generateValue() {
         var type = randomFrom(ShapeType.values());
         var isGeoJson = randomBoolean();
 
-        return switch (type) {
-            // LINEARRING and CIRCLE are not supported as inputs to fields so just return points
-            case POINT, LINEARRING, CIRCLE -> {
-                var point = GeometryTestUtils.randomPoint(false);
-                yield value(point, isGeoJson);
+        while (true) {
+            Geometry candidateGeometry = switch (type) {
+                // LINEARRING and CIRCLE are not supported as inputs to fields so just return points
+                case POINT, LINEARRING, CIRCLE -> GeometryTestUtils.randomPoint(false);
+                case MULTIPOINT -> GeometryTestUtils.randomMultiPoint(false);
+                case LINESTRING -> GeometryTestUtils.randomLine(false);
+                case MULTILINESTRING -> GeometryTestUtils.randomMultiLine(false);
+                case POLYGON -> GeometryTestUtils.randomPolygon(false);
+                case MULTIPOLYGON -> GeometryTestUtils.randomMultiPolygon(false);
+                case GEOMETRYCOLLECTION -> GeometryTestUtils.randomGeometryCollectionWithoutCircle(false);
+                case ENVELOPE -> GeometryTestUtils.randomRectangle();
+            };
+
+            try {
+                if (fieldType == FieldType.GEO_SHAPE) {
+                    GeoTestUtils.binaryGeoShapeDocValuesField("f", candidateGeometry);
+                } else {
+                    GeoTestUtils.binaryCartesianShapeDocValuesField("f", candidateGeometry);
+                }
+
+                if (type == ShapeType.ENVELOPE) {
+                    var wktString = WellKnownText.toWKT(candidateGeometry);
+
+                    return new Value(wktString, wktString);
+                }
+
+                return value(candidateGeometry, isGeoJson);
+            } catch (IllegalArgumentException ignored) {
+                // It's malformed somehow, loop
```

Comment:
I didn't get this part, why do we loop here?

Reply:
`GeometryTestUtils.random*` sometimes produces invalid shapes and we are not interested in those here. Loop will regenerate shape of the same type.